### ISSUE TITLE
ensured npm update 

### ISSUE THREAD

**Issue by Spown** [icflorescu/openshift-cartridge-nodejs#42]
so I was playing with autoupdating node and npm versions through autogenerated from package.json env variables and at some point it installed node 4 and npm 1.1.37 (still don't get how) and aftarwards no further updating was possible and the deployment failed every time. Even when I corrected my `NPM_VERSION_URL` it just failed because apperantly [you are using npm to update itself](https://github.com/icflorescu/openshift-cartridge-nodejs/blob/master/lib/util#L58)
![ ](http://www.reactionface.info/sites/default/files/imagecache/Node_Page/images/1257003814239.jpg)

so basically if it gets broken and fails to start once - it's lights out forever. Well fortunately in my case it wasn't really broken - it was just incompatable with node, so I had to fall back to node@0.8, update to npm@~2.14, update back to node@~4.4, reinstall global modules manually... good times... But imagine if npm gets really broken for some reason. The whole gear dies.  

Maybe it would be wiser to use some other updating method? maybe a sandboxed node+npm pack dedicated solely to updating npm? Or just like the node - through a dist archive?

**Comment by icflorescu** [comment]
I really have no idea what could have happened for your gear to end up with Node.js 4 & npm 1.1.37. There are no such defaults in the cartridge. Maybe a temporary bug in the npm registry? (it happened before)... But that would not explain the Node.js version. As about using npm to update itself, it's common practice AFAIK.

![keep calm](https://cloud.githubusercontent.com/assets/581999/15635600/66e6a906-25eb-11e6-8ce8-649ac4982d31.jpg)

Of course, there's always a minimal risk of ending up with a broken instance, that's why it's a good idea to backup your data :-). If we're talking about a critical application, there are safer options such as paid plans with docker-based deployments or even other PaaS providers such as Heroku (a lot easier to deploy) or Nitrous (which is even more dev-oriented than OpenShift - I think they'll even grant you root access on the box).

I'd be reluctant to implement a different and more complicated install/update method. The `update_nodejs` function got complicated enough, and adding more logic to the whole workflow won't make it faster or safer. If we keep adding more and more features we'll soon end up with a cartridge taking half an hour to spin up, or rather with a 50% chance of a timed-out deployment...

I recognize my approach won't accommodate everyone's requirements and particular needs, but in the end, this is a community cartridge, I'm in no way affiliated with RedHat, so, by all means, feel free to fork it and PR ;-).

**Comment by Spown** [comment]
that's a very long way to say "I won't do anything"...

> As about using npm to update itself, it's common practice AFAIK.

when you are doing it manually, yes, but not in an automated process.

> I'd be reluctant to implement a different and more complicated install/update method.
> I recognize my approach won't accommodate everyone's requirements and particular needs, but in the end, this is a community cartridge, I'm in no way affiliated with RedHat, so, by all means, feel free to fork it and PR ;-).

Well, I'm not demanding support or anything - as I said I managed to solve my problem myself. I'm just pointing at an obvious flaw. Increasing the code difficulty is almost always means increasing the enthropy and everybody has their own view on the complexity/stability ballance. But if you ask me, a more complex function that relies on tar in the end is better than a simpler one that relies on a much more prune to fail mechanic... especially if every future call of the function relies on the succesful result of it's run... It's jus a suggestion.

**Comment by icflorescu** [comment]
I'm sorry, I was aiming for the _polite_ way to say "I don't have time to really look into it at the moment" :-)

Truth is, I've became a bit frustrated lately over this repo. While GitHub says there are constantly ~250 clones per day (= as many deployments), there were only [5 other contributors](https://github.com/icflorescu/openshift-cartridge-nodejs/graphs/contributors) so far, but I'm constantly getting emails from people looking for free support and advice. I mean, I'm glad the cartridge is useful to so many, but I wouldn't have the time & energy to accommodate all their wishes (not to mention the obligation) and I'm sad to see so few are understanding what open-source truly means.

Btw, that's the polite and long way of saying "I'm looking for committing contributors". Deployments on OpenShift aren't exactly blazing fast and testing new features takes quite a while.

**Comment by icflorescu** [comment]
Going back to the matter of `npm i -g npm` being ok _"when you are doing it manually, yes, but not in an automated process"_, can't argue with that, but it's the only way I've seen it done, even on [Heroku's Node.js buildpack](https://github.com/heroku/heroku-buildpack-nodejs/blob/master/lib/binaries.sh#L59).

**Comment by Spown** [comment]
As I managed to stabilize my current project I don't really need a custom custom cartridge right now, however I have an idea of speeding up build process by updating only those modules that need to be updated (via [npm outdated](https://docs.npmjs.com/cli/outdated)) and that would require forking your project, so when/if I get there I can also introduce an alternative npm updater. This whole thing however is pretty low priority on my project plan...

**Comment by icflorescu** [comment]
Good to know. When/if you get there, I'll have a look at it, but it's pretty low priority on my plan too.
Also, make sure to have a look at OpenShift's cartridge dev guide. Right now I'm already speeding up the build process by using OpenShift's `pre-repo-archive` and `build` events. Npms are cached before deployment; when you deploy, the unneeded ones are pruned and just the outdated ones are installed.

**Comment by icflorescu** [comment]
Going to close this as non-constructive for now. Auto-updating `npm` is the _de-facto_ practice among PaaS providers and reinventing the mechanism would be counter-productive.
